### PR TITLE
feature(generalization): Make the Allen metric usable with alternate source models

### PR DESCRIPTION
**Pull request by rju**
Introduced minor changes to allow to use different Node implementations. Node is now an interface. This allows to use different Node implementation with the metric to allow the use of different source models for the graph.

### CODE REVIEW

**Comment by layornos** [comment]
Please run `gradle spotlessApply` before committing to avoid formatting changes and to comply to the desired formatting style

**Comment by rju** [comment]
I did run ./gradlew spotlessApply it does not produce any errors.

**Comment by rju** [comment]
I'll check whether something else is broken.

**Comment by rju** [comment]
The issue seems to be that it cannot perform actions/checkout@v2
Producing the following error: "Input required and not supplied: token"
Looks like the access token is missing for some strange reason.

**Review by MartinWitt — APPROVED** [review]

**Review by MartinWitt — APPROVED** [review]

**Comment by MartinWitt** [comment]
Hey, 
this issue seems related to some strange GitHub security concept. In #113 I remove the old unused token and now PRs should run actions. Can you merge the current master in your branch or set the checkbox, that I'm allowed to push commits to your branch? 
The checkbox should look something like this.
image

File: architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/SystemGraphUtils.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+/** */
+package edu.kit.kastel.sdq.case4lang.refactorlizar.architecture_evaluation.graphs;
+
+import com.google.common.graph.Graph;
+import com.google.common.graph.MutableGraph;
+
+/** @author Reiner Jung */
```
Comment (MartinWitt) [inline architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/SystemGraphUtils.java:7]: warning: [MissingSummary] A summary line is required on public/protected Javadocs.
/** @author Reiner Jung */
    ^
    (see https://google.github.io/styleguide/javaguide.html#s7.2-summary-fragment)

File: architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/Node.java [hunk 2]
```diff
@@ -1,53 +1,10 @@
...
 
-import java.util.Objects;
-import spoon.reflect.declaration.CtType;
-import spoon.reflect.declaration.CtTypeMember;
-import spoon.reflect.visitor.CtVisitor;
+/**
+ * @author Reiner Jung
+ * @param <T> module identifying type
+ */
+public interface Node<T> {
```
Comment (MartinWitt) [inline architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/Node.java:7]: warning: [MissingSummary] A summary line is required on public/protected Javadocs.
 * @author Reiner Jung
   ^
    (see https://google.github.io/styleguide/javaguide.html#s7.2-summary-fragment)

File: architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/CtTypeSystemGraphUtils.java [hunk 3]
```diff
@@ -6,17 +6,17 @@
...
+    public MutableGraph<Node<CtType<?>>> convertToSystemGraph(Graph<Node<CtType<?>>> graph) {
+        MutableGraph<Node<CtType<?>>> systemGraph = Graphs.copyOf(graph);
         // add empty node as system node
         CtType<?> type = Launcher.parseClass("class ThIsIsRiskant {}");
         var method = new Launcher().createFactory().createMethod();
         type.addMethod(method);
-        systemGraph.addNode(new Node(method));
+        systemGraph.addNode(new SpoonNode(method));
         return systemGraph;
     }
```
Comment (MartinWitt) [inline architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/CtTypeSystemGraphUtils.java:21]: warning: [MissingOverride] convertToSystemGraph implements method in SystemGraphUtils; expected @Override
    public MutableGraph<Node<CtType<?>>> convertToSystemGraph(Graph<Node<CtType<?>>> graph) {

File: architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/SpoonNode.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+    /** @param member */
+    public SpoonNode(CtTypeMember member) {
+        this.member = member;
+    }
+
+    /* (non-Javadoc)
+     * @see java.lang.Object#hashCode()
+     */
+
```
Comment (MartinWitt) [inline architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/SpoonNode.java:19]: warning: [AlmostJavadoc] This comment contains Javadoc or HTML tags, but isn't started with a double asterisk (/**); is it meant to be Javadoc?

File: architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/SpoonNode.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) return true;
+        if (!(obj instanceof SpoonNode)) return false;
+        SpoonNode other = (SpoonNode) obj;
+        return Objects.equals(member, other.member)
+                && Objects.equals(member.getDeclaringType(), other.member.getDeclaringType());
+    }
```
Comment (MartinWitt) [inline architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/SpoonNode.java:36]: warning: [AlmostJavadoc] This comment contains Javadoc or HTML tags, but isn't started with a double asterisk (/**); is it meant to be Javadoc?

File: architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/SpoonNode.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        member.accept(arg0);
+    }
+
+    /**
+     * @return
+     * @see spoon.reflect.declaration.CtTypeMember#getDeclaringType()
+     */
+    public CtType<?> getModule() {
+        return member.getDeclaringType();
+    }
```
Comment (MartinWitt) [inline architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/SpoonNode.java:52]: [EmptyBlockTag] A block tag (@param, @return, @throws, @deprecated) has an empty description. Block tags without descriptions don't add much value for future readers of the code; consider removing the tag entirely or adding a description.

[MissingSummary] A summary fragment is required; consider using the value of the @see block as a summary fragment instead.

[MissingOverride] getModule implements method in Node; expected @Override
    public CtType<?> getModule() {

File: architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/SpoonNode.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+                && Objects.equals(member.getDeclaringType(), other.member.getDeclaringType());
+    }
+
+    /**
+     * @param arg0
+     * @see spoon.reflect.visitor.CtVisitable#accept(spoon.reflect.visitor.CtVisitor)
+     */
+    public void accept(CtVisitor arg0) {
+        member.accept(arg0);
+    }
```
Comment (MartinWitt) [inline architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/SpoonNode.java:44]: see method below

**Review by MartinWitt — CHANGES_REQUESTED** [review]
I have added the comments from errorprone. These are minor style warnings. Could you fix them?

**Comment by layornos** [comment]
> I did run ./gradlew spotlessApply it does not produce any errors.

I've seen many changed files with only style changes, maybe we've forgot to run spotless? idk

**Comment by MartinWitt** [comment]
@rju sry, I forgot we had 3 workflow files and all had the token usage, could you merge again? I can fix the commit history of your PR later, don't worry.

File: architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/SystemGraphUtils.java (hunk 1 above)
Comment (rju) [reply architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/SystemGraphUtils.java:7]: Added a comment.

File: architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/Node.java (hunk 2 above)
Comment (rju) [reply architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/Node.java:7]: Also added a comment.

File: architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/CtTypeSystemGraphUtils.java (hunk 3 above)
Comment (rju) [reply architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/CtTypeSystemGraphUtils.java:21]: It shows that I did these changes with an unconfigured Eclipse workspace (for Kieker we have instructions for that :-)). Added the @override.

File: architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/SpoonNode.java (hunk 5 above)
Comment (rju) [reply architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/SpoonNode.java:36]: This was just the comment from the original class. I did not change that.

File: architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/SpoonNode.java (hunk 6 above)
Comment (rju) [reply architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/SpoonNode.java]: Also I did not change this method, just renamed it. Anyway, added a comment.

File: architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/SpoonNode.java (hunk 7 above)
Comment (rju) [reply architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/SpoonNode.java]: Added a comment, but the code looked like this before I added code.

**Comment by rju** [comment]
Added comments and overrides where necessary. The spotlessCheck and apply feature is neat. I will add this to Kieker.

File: architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/SpoonNode.java (hunk 6 above)
Comment (MartinWitt) [reply architecture-evaluation/src/main/java/edu/kit/kastel/sdq/case4lang/refactorlizar/architecture_evaluation/graphs/SpoonNode.java]: Sadly, there were already some error-prone smells, I tried to filter them. We still try to reduce technical debt, there is not always time left for it. Thanks for the comment and sorry for the inconvenience.

**Comment by MartinWitt** [comment]
@layornos PR looks good from my side and ready for squash+merge.

**Review by MartinWitt — APPROVED** [review]

**Pull request merged**